src/condor_utils/file_transfer.cpp

```diff
@@ -6435,15 +6468,68 @@ FileTransfer::ExpandFileTransferList( char const *src_path, char const *dest_dir
 			// dprintf( D_ALWAYS, ">>> not preserving relative path.\n" );
 			destination += condor_basename(src_path);
 		} else {
-			// dprintf( D_ALWAYS, ">>> preserving relative path.\n" );
-			destination += src_path;
+			if(! fullpath(src_path)) {
+				// dprintf( D_ALWAYS, ">>> preserving relative path of relative path %s\n", src_path );
 
-			// ExpandParentDirectories() adds this back in the correct place.
-			expanded_list.pop_back();
+				if( destination.length() > 0 ) { destination += DIR_DELIM_CHAR; }
+				destination += src_path;
+
+				// ExpandParentDirectories() adds this back in the correct place.
+				expanded_list.pop_back();
+
+				if(! ExpandParentDirectories( src_path, iwd, expanded_list, SpoolSpace )) {
+					return false;
+				}
+			} else {
+				//
+				// The only absolute paths we want to treat as relative paths
+				// are absolute paths into the SPOOL directory.  Everything
+				// else should be transferred as usual.
+				//
+
+				ASSERT( SpoolSpace == NULL || fullpath(SpoolSpace) );
```

Comment:
An `ASSERT` seems harsh here.  Why not just fail?

Reply:
Because if `SpoolSpace` is a relative path, then something has gone terribly wrong.

We also don't check the the return value of `ExpandFileTransferList()` in  its root call from `DoUpload()`, so returning false presently won't actually prevent the transfer happening, so this was the smaller code change.

Reply:
Can you at least leave a comment stating that `SpoolSpace` is not under user control; we assume if the admin screws up this setting than it's OK to assert and die.

Reply:
Done.